File: sim_rk_test/src/rk3568_udp_communication.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
RK3568与FPGA UDP通信程序
FPGA端配置:
  - FPGA IP: 192.168.1.110
  - FPGA端口: 8080
  - RK3568 IP: 192.168.1.105
  - RK3568端口: 8080
"""

import socket
import threading
import time
from datetime import datetime

class RK3568_UDP_Communication:
    def __init__(self, local_ip='192.168.137.116', local_port=8080, 
                 fpga_ip='192.168.137.1', fpga_port=8080):
        """
        初始化UDP通信
        
        Args:
            local_ip: 本机IP地址
            local_port: 本机监听端口
            fpga_ip: FPGA的IP地址
            fpga_port: FPGA的端口
        """
        self.local_ip = local_ip
        self.local_port = local_port
        self.fpga_ip = fpga_ip
        self.fpga_port = fpga_port
        
        # 创建UDP socket
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        
        # 绑定本地地址和端口
        self.sock.bind((self.local_ip, self.local_port))
        
        # 设置超时，避免阻塞
        self.sock.settimeout(1.0)
        
        self.running = False
        self.receive_thread = None
        
        print(f"[INFO] UDP Socket初始化成功")
        print(f"[INFO] 本机地址: {self.local_ip}:{self.local_port}")
        print(f"[INFO] FPGA地址: {self.fpga_ip}:{self.fpga_port}")
    
    def start_receive_thread(self):
        """启动接收线程"""
        self.running = True
        self.receive_thread = threading.Thread(target=self._receive_loop, daemon=True)
        self.receive_thread.start()
        print("[INFO] 接收线程已启动")
    
    def _receive_loop(self):
        """接收数据循环"""
        while self.running:
            try:
                # 接收数据
                data, addr = self.sock.recvfrom(1024)
                timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
                
                print(f"\n[{timestamp}] 收到来自 {addr[0]}:{addr[1]} 的数据")
                print(f"[RECV] 数据长度: {len(data)} 字节")
                print(f"[RECV] 数据内容 (hex): {data.hex()}")
                print(f"[RECV] 数据内容 (ascii): {data.decode('ascii', errors='replace')}")
                print(f"[RECV] 数据内容 (bytes): {list(data)}")
                
            except socket.timeout:
                # 超时是正常的，继续循环
                continue
            except Exception as e:
                print(f"[ERROR] 接收数据错误: {e}")
                time.sleep(0.1)
    
    def send_data(self, data):
        """
        发送数据到FPGA
        
        Args:
            data: 要发送的数据，可以是字符串或字节
        """
        try:
            if isinstance(data, str):
                data = data.encode('utf-8')
            
            self.sock.sendto(data, (self.fpga_ip, self.fpga_port))
            timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
            print(f"[{timestamp}] 发送数据到 {self.fpga_ip}:{self.fpga_port}")
            print(f"[SEND] 数据长度: {len(data)} 字节")
            print(f"[SEND] 数据内容 (hex): {data.hex()}")
            return True
        except Exception as e:
            print(f"[ERROR] 发送数据错误: {e}")
            return False
    
    def send_command(self, command):
        """
        发送命令到FPGA
        
        Args:
            command: 命令字符串
        """
        return self.send_data(command)
    
    def stop(self):
        """停止通信"""
        self.running = False
        if self.receive_thread:
            self.receive_thread.join(timeout=2)
        self.sock.close()
        print("[INFO] UDP通信已停止")


def main():
    """主函数 - 交互式测试"""
    print("=" * 60)
    print("RK3568与FPGA UDP通信测试程序")
    print("=" * 60)
    
    # 创建通信对象
    udp_comm = RK3568_UDP_Communication()
    
    # 启动接收线程
    udp_comm.start_receive_thread()
    
    print("\n[提示] 等待FPGA发送数据...")
    print("[提示] 输入 'help' 查看命令列表")
    print("-" * 60)
    
    try:
        while True:
            # 获取用户输入
            user_input = input("\n请输入要发送的数据 (或输入 'quit' 退出): ").strip()
            
            if user_input.lower() == 'quit':
                print("[INFO] 正在退出程序...")
                break
            elif user_input.lower() == 'help':
                print("\n命令列表:")
                print("  help     - 显示帮助信息")
                print("  quit     - 退出程序")
                print("  test     - 发送测试数据")
                print("  <数据>   - 发送任意数据到FPGA")
                print("\n当前状态:")
                print(f"  本机: {udp_comm.local_ip}:{udp_comm.local_port}")
                print(f"  FPGA: {udp_comm.fpga_ip}:{udp_comm.fpga_port}")
            elif user_input.lower() == 'test':
                # 发送测试数据
                test_data = "RK3568_TEST_DATA"
                udp_comm.send_data(test_data)
            elif user_input:
                # 发送用户输入的数据
                udp_comm.send_data(user_input)
    
    except KeyboardInterrupt:
        print("\n[INFO] 收到中断信号，正在退出...")
    finally:
        udp_comm.stop()


if __name__ == "__main__":
    main()

File: sim_rk_test/src/rk3568_udp_communication.c
/**
 * RK3568与FPGA UDP通信程序 (C语言版本)
 * 
 * 编译命令:
 *   gcc rk3568_udp_communication.c -o rk3568_udp
 * 
 * 运行:
 *   ./rk3568_udp
 * 
 * FPGA端配置:
 *   - FPGA IP: 192.168.1.110
 *   - FPGA端口: 8080
 *   - RK3568 IP: 192.168.1.105
 *   - RK3568端口: 8080
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <time.h>
#include <signal.h>
#include <pthread.h>

#define BUFFER_SIZE 1024
#define LOCAL_IP "192.168.1.105"
#define LOCAL_PORT 8080
#define FPGA_IP "192.168.1.110"
#define FPGA_PORT 8080

// 全局变量
static int running = 1;
static int sock_fd = -1;

/**
 * 获取当前时间戳字符串
 */
void get_timestamp(char *buffer, size_t size) {
    struct timespec ts;
    struct tm *tm_info;
    
    clock_gettime(CLOCK_REALTIME, &ts);
    tm_info = localtime(&ts.tv_sec);
    
    snprintf(buffer, size, "%04d-%02d-%02d %02d:%02d:%02d.%03ld",
             tm_info->tm_year + 1900, tm_info->tm_mon + 1, tm_info->tm_mday,
             tm_info->tm_hour, tm_info->tm_min, tm_info->tm_sec,
             ts.tv_nsec / 1000000);
}

/**
 * 信号处理函数
 */
void signal_handler(int sig) {
    printf("\n[INFO] 收到信号 %d，正在退出...\n", sig);
    running = 0;
}

/**
 * 打印数据内容（多种格式）
 */
void print_data(const char *prefix, const unsigned char *data, int len) {
    char timestamp[32];
    get_timestamp(timestamp, sizeof(timestamp));
    
    printf("[%s] %s 数据长度: %d 字节\n", timestamp, prefix, len);
    printf("[%s] %s 数据内容 (hex): ", timestamp, prefix);
    for (int i = 0; i < len && i < 32; i++) {
        printf("%02X ", data[i]);
    }
    if (len > 32) printf("...");
    printf("\n");
    
    printf("[%s] %s 数据内容 (ascii): ", timestamp, prefix);
    for (int i = 0; i < len && i < 64; i++) {
        if (data[i] >= 32 && data[i] <= 126) {
            printf("%c", data[i]);
        } else {
            printf(".");
        }
    }
    if (len > 64) printf("...");
    printf("\n");
}

/**
 * 接收线程函数
 */
void *receive_thread_func(void *arg) {
    unsigned char buffer[BUFFER_SIZE];
    struct sockaddr_in from_addr;
    socklen_t from_len = sizeof(from_addr);
    
    printf("[INFO] 接收线程已启动\n");
    
    while (running) {
        // 设置超时
        struct timeval timeout;
        timeout.tv_sec = 1;
        timeout.tv_usec = 0;
        
        if (setsockopt(sock_fd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout)) < 0) {
            perror("[ERROR] 设置接收超时失败");
            break;
        }
        
        // 接收数据
        int recv_len = recvfrom(sock_fd, buffer, BUFFER_SIZE - 1, 0,
                               (struct sockaddr *)&from_addr, &from_len);
        
        if (recv_len < 0) {
            if (errno == EAGAIN || errno == EWOULDBLOCK) {
                // 超时是正常的，继续循环
                continue;
            }
            perror("[ERROR] 接收数据失败");
            break;
        }
        
        // 打印接收到的数据
        char timestamp[32];
        get_timestamp(timestamp, sizeof(timestamp));
        printf("\n[%s] 收到来自 %s:%d 的数据\n", timestamp,
               inet_ntoa(from_addr.sin_addr), ntohs(from_addr.sin_port));
        
        print_data("RECV", buffer, recv_len);
        printf("\n");
    }
    
    printf("[INFO] 接收线程已退出\n");
    return NULL;
}

/**
 * 发送数据到FPGA
 */
int send_to_fpga(const char *data, int len) {
    struct sockaddr_in fpga_addr;
    
    memset(&fpga_addr, 0, sizeof(fpga_addr));
    fpga_addr.sin_family = AF_INET;
    fpga_addr.sin_port = htons(FPGA_PORT);
    
    if (inet_pton(AF_INET, FPGA_IP, &fpga_addr.sin_addr) <= 0) {
        perror("[ERROR] 无效的FPGA IP地址");
        return -1;
    }
    
    int sent_len = sendto(sock_fd, data, len, 0,
                         (struct sockaddr *)&fpga_addr, sizeof(fpga_addr));
    
    if (sent_len < 0) {
        perror("[ERROR] 发送数据失败");
        return -1;
    }
    
    char timestamp[32];
    get_timestamp(timestamp, sizeof(timestamp));
    printf("[%s] 发送数据到 %s:%d\n", timestamp, FPGA_IP, FPGA_PORT);
    print_data("SEND", (unsigned char *)data, sent_len);
    
    return sent_len;
}

/**
 * 主函数
 */
int main(int argc, char *argv[]) {
    int ret;
    pthread_t recv_thread;
    struct sockaddr_in local_addr;
    
    printf("========================================\n");
    printf("RK3568与FPGA UDP通信测试程序 (C版本)\n");
    printf("========================================\n");
    
    // 注册信号处理
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    
    // 创建UDP socket
    sock_fd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock_fd < 0) {
        perror("[ERROR] 创建socket失败");
        return -1;
    }
    
    // 设置地址重用
    int reuse = 1;
    if (setsockopt(sock_fd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse)) < 0) {
        perror("[ERROR] 设置地址重用失败");
        close(sock_fd);
        return -1;
    }
    
    // 绑定本地地址
    memset(&local_addr, 0, sizeof(local_addr));
    local_addr.sin_family = AF_INET;
    local_addr.sin_port = htons(LOCAL_PORT);
    local_addr.sin_addr.s_addr = inet_addr(LOCAL_IP);
    
    if (bind(sock_fd, (struct sockaddr *)&local_addr, sizeof(local_addr)) < 0) {
        perror("[ERROR] 绑定地址失败");
        close(sock_fd);
        return -1;
    }
    
    printf("[INFO] UDP Socket初始化成功\n");
    printf("[INFO] 本机地址: %s:%d\n", LOCAL_IP, LOCAL_PORT);
    printf("[INFO] FPGA地址: %s:%d\n", FPGA_IP, FPGA_PORT);
    
    // 启动接收线程
    ret = pthread_create(&recv_thread, NULL, receive_thread_func, NULL);
    if (ret != 0) {
        printf("[ERROR] 创建接收线程失败: %s\n", strerror(ret));
        close(sock_fd);
        return -1;
    }
    
    printf("\n[提示] 等待FPGA发送数据...\n");
    printf("[提示] 输入 'help' 查看命令列表\n");
    printf("----------------------------------------\n");
    
    // 主循环 - 处理用户输入
    char input[BUFFER_SIZE];
    while (running) {
        printf("\n请输入要发送的数据 (或输入 'quit' 退出): ");
        fflush(stdout);
        
        if (fgets(input, sizeof(input), stdin) == NULL) {
            break;
        }
        
        // 去除换行符
        input[strcspn(input, "\n")] = 0;
        
        if (strlen(input) == 0) {
            continue;
        }
        
        if (strcmp(input, "quit") == 0) {
            printf("[INFO] 正在退出程序...\n");
            break;
        } else if (strcmp(input, "help") == 0) {
            printf("\n命令列表:\n");
            printf("  help     - 显示帮助信息\n");
            printf("  quit     - 退出程序\n");
            printf("  test     - 发送测试数据\n");
            printf("  <数据>   - 发送任意数据到FPGA\n");
            printf("\n当前状态:\n");
            printf("  本机: %s:%d\n", LOCAL_IP, LOCAL_PORT);
            printf("  FPGA: %s:%d\n", FPGA_IP, FPGA_PORT);
        } else if (strcmp(input, "test") == 0) {
            // 发送测试数据
            const char *test_data = "RK3568_TEST_DATA";
            send_to_fpga(test_data, strlen(test_data));
        } else {
            // 发送用户输入的数据
            send_to_fpga(input, strlen(input));
        }
    }
    
    // 清理资源
    running = 0;
    pthread_join(recv_thread, NULL);
    close(sock_fd);
    
    printf("[INFO] UDP通信已停止\n");
    return 0;
}